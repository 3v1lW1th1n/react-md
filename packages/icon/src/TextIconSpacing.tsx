import * as React from "react";
import * as PropTypes from "prop-types";
import cn from "classnames";

export interface ITextIconSpacingProps {
  /**
   * An optional className to apply to the surroudning `<span>` when the `forceIconWrap` prop is enabled
   * or the icon is not a valid React Element.
   *
   * @docgen
   */
  className?: string;

  /**
   * An optional icon to display with a text button. This is invalid for icon buttons. If this is
   * a single element, a new class name will be cloned into the element to get correct spacing so
   * if you have a custom icon element, you **must** also pass that class name down. If you are using
   * one of the react-md icon component packages, this is handled automatically.
   *
   * If this is not a valid react element, the icon will be wrapped in a `<span>` instead
   * with the class names applied.
   *
   * @docgen
   */
  icon?: React.ReactElement<Element> | React.ReactNode;

  /**
   * Boolean if the icon should appear after the text instead of before.
   *
   * @docgen
   */
  iconAfter?: boolean;

  /**
   * The children to render before or after the provided icon.
   *
   * @docgen
   */
  children?: React.ReactNode;

  /**
   * The class name to use for an icon that is placed before text.
   *
   * @docgen
   */
  beforeClassName?: string;

  /**
   * The class name to use for an icon that is placed after text.
   *
   * @docgen
   */
  afterClassName?: string;

  /**
   * Boolean if the icon should be forced into a `<span>` with the class names applied instead of attempting
   * to clone into the provided icon.
   *
   * @docgen
   */
  forceIconWrap?: boolean;
}

export interface ITextIconSpacingDefaultProps {
  iconAfter: boolean;
  beforeClassName: string;
  afterClassName: string;
  forceIconWrap: boolean;
}

export type TextIconSpacingWithDefaultProps = ITextIconSpacingProps & ITextIconSpacingDefaultProps;

export default class TextIconSpacing extends React.Component<ITextIconSpacingProps, {}> {
  public static propTypes = {
    icon: PropTypes.oneOfType([PropTypes.element, PropTypes.node]),
    iconAfter: PropTypes.bool,
    beforeClassName: PropTypes.string,
    afterClassName: PropTypes.string,
    forceIconWrap: PropTypes.bool,
    children: PropTypes.node,
  };

  public static defaultProps: ITextIconSpacingDefaultProps = {
    iconAfter: false,
    forceIconWrap: false,
    beforeClassName: "rmd-icon--before",
    afterClassName: "rmd-icon--after",
  };

  public render() {
    const {
      icon: propIcon,
      iconAfter,
      children,
      className,
      beforeClassName,
      afterClassName,
      forceIconWrap,
    } = this.props as TextIconSpacingWithDefaultProps;

    if (!propIcon) {
      return children;
    }

    let iconEl = propIcon;
    let content = children;
    if (!forceIconWrap && React.isValidElement(propIcon)) {
      const icon = React.Children.only(propIcon);
      iconEl = React.cloneElement(icon, {
        className: cn(
          {
            [beforeClassName]: !iconAfter,
            [afterClassName]: iconAfter,
          },
          icon.props.className
        ),
      });
    } else if (propIcon) {
      iconEl = (
        <span
          className={cn("rmd-text-icon-spacing", {
            [beforeClassName]: !iconAfter,
            [afterClassName]: iconAfter,
          }, className)}
        >
          {propIcon}
        </span>
      );
    }

    if (iconEl) {
      content = (
        <React.Fragment>
          {!iconAfter && iconEl}
          {children}
          {iconAfter && iconEl}
        </React.Fragment>
      );
    }

    return content;
  }
}
