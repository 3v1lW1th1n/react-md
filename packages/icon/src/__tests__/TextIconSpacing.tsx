import * as React from "react";
import { create } from "react-test-renderer";
import { mount } from "enzyme";

import TextIconSpacing from "../TextIconSpacing";

describe("TextIconSpacing", () => {
  it("should be able to render without any children or icon even though it's worthless", () => {
    expect(create(<TextIconSpacing />).toJSON()).toMatchSnapshot();
  });

  it("should return the children if no icon prop is provided", () => {
    const instance = mount(
      <TextIconSpacing>
        <span />
      </TextIconSpacing>
    );
    expect(instance.render()).toMatchSnapshot();
    expect(instance.find("span").length).toBe(1);
    expect(instance.find("span").props().className).toBeUndefined();
  });

  it("should render the icon before the children by default", () => {
    const instance = mount(
      <TextIconSpacing icon={<i />}>
        <span />
      </TextIconSpacing>
    );

    expect(
      instance
        .children()
        .first()
        .type()
    ).toBe("i");
    expect(
      instance
        .children()
        .at(1)
        .type()
    ).toBe("span");
  });

  it("should render the icon after the children if the iconAfter prop is enabled", () => {
    const instance = mount(
      <TextIconSpacing icon={<i />} iconAfter>
        <span />
      </TextIconSpacing>
    );

    expect(
      instance
        .children()
        .first()
        .type()
    ).toBe("span");
    expect(
      instance
        .children()
        .at(1)
        .type()
    ).toBe("i");
  });

  it("should clone the default class names into the icon element", () => {
    const instance = mount(
      <TextIconSpacing icon={<i />}>
        <span />
      </TextIconSpacing>
    );

    expect(instance.find("i").hasClass("rmd-icon--before")).toBe(true);
    expect(instance.find("i").hasClass("rmd-icon--after")).toBe(false);

    instance.setProps({ iconAfter: true });
    expect(instance.find("i").hasClass("rmd-icon--before")).toBe(false);
    expect(instance.find("i").hasClass("rmd-icon--after")).toBe(true);
  });

  it("should wrap the icon in a span with the required classNames if the icon is not a valid react element or the forceIconWrap prop is enabled", () => {
    // have to wrap the tests in divs since enzyme doesn't know how to render React.Fragment or arrays atm
    const instance1 = mount(
      <div>
        <TextIconSpacing icon="Some text">
          <div />
        </TextIconSpacing>
      </div>
    );

    expect(instance1.render()).toMatchSnapshot();
    expect(instance1.find("span").hasClass("rmd-text-icon-spacing")).toBe(true);
    expect(instance1.find("span").hasClass("rmd-icon--before")).toBe(true);

    const instance2 = mount(
      <div>
        <TextIconSpacing icon={<i />} forceIconWrap>
          <div />
        </TextIconSpacing>
      </div>
    );

    expect(instance2.render()).toMatchSnapshot();
    expect(instance2.find("span").hasClass("rmd-text-icon-spacing")).toBe(true);
    expect(instance2.find("span").hasClass("rmd-icon--before")).toBe(true);
  });
});
