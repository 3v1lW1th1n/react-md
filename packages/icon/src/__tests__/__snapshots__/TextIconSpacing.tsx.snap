// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextIconSpacing should be able to render without any children or icon even though it's worthless 1`] = `null`;

exports[`TextIconSpacing should return the children if no icon prop is provided 1`] = `<span />`;

exports[`TextIconSpacing should wrap the icon in a span with the required classNames if the icon is not a valid react element or the forceIconWrap prop is enabled 1`] = `
<div>
  <span
    class="rmd-text-icon-spacing rmd-icon--before"
  >
    Some text
  </span>
  <div />
</div>
`;

exports[`TextIconSpacing should wrap the icon in a span with the required classNames if the icon is not a valid react element or the forceIconWrap prop is enabled 2`] = `
<div>
  <span
    class="rmd-text-icon-spacing rmd-icon--before"
  >
    <i />
  </span>
  <div />
</div>
`;
